tests/test_Restart.cpp

```diff
@@ -0,0 +1,300 @@
+
+    sol.insert( ds::RS, { rsi.begin(), rsi.end() } );
+    sol.insert( ds::RV, { rvi.begin(), rvi.end() } );
+
+    return sol;
+}
+
+std::pair< data::Solution, data::Wells >
+first_sim(test_work_area_type * test_area) {
+
+    std::string eclipse_data_filename    = "FIRST_SIM.DATA";
+    test_work_area_copy_file(test_area, eclipse_data_filename.c_str());
+
+    auto eclipseState = std::make_shared< EclipseState >(
+            Parser::parse( eclipse_data_filename ) );
+
+    const auto& grid = *eclipseState->getInputGrid();
+    auto num_cells = grid.getNX() * grid.getNY() * grid.getNZ();
+
+    EclipseWriter eclWriter( eclipseState, num_cells, nullptr );
+    auto start_time = util_make_datetime( 0, 0, 0, 1, 11, 1979 );
+    auto first_step = util_make_datetime( 0, 0, 0, 10, 10, 2008 );
+    eclWriter.writeInit( NNC() );
+
+    auto sol = mkSolution( num_cells );
+    auto wells = mkWells();
+
+    eclWriter.writeTimeStep( 1,
+                             first_step - start_time,
+                             sol, wells, false);
+
+    return { sol, wells };
+}
+
+std::pair< data::Solution, data::Wells > second_sim() {
+    auto eclipseState = Parser::parseData( input );
+
+    const auto& grid = *eclipseState.getInputGrid();
+    auto num_cells = grid.getNX() * grid.getNY() * grid.getNZ();
+
+    return init_from_restart_file( eclipseState, num_cells );
+}
+
+void compare( std::pair< data::Solution, data::Wells > fst,
+              std::pair< data::Solution, data::Wells > snd ) {
+
+    using ds = data::Solution::key;
+    for( auto key : { ds::PRESSURE, ds::TEMP, ds::SWAT, ds::SGAS,
+                      ds::RS, ds::RV } ) {
+
+        auto first = fst.first[ key ].begin();
+        auto second = snd.first[ key ].begin();
+
+        for( ; first != fst.first[ key ].end(); ++first, ++second )
+            BOOST_CHECK_CLOSE( *first, *second, 0.00001 );
+    }
+
+    BOOST_CHECK_EQUAL_COLLECTIONS(
+            fst.second.bhp.begin(), fst.second.bhp.end(),
+            snd.second.bhp.begin(), snd.second.bhp.end()
+        );
+    BOOST_CHECK_EQUAL_COLLECTIONS(
+            fst.second.temperature.begin(), fst.second.temperature.end(),
+            snd.second.temperature.begin(), snd.second.temperature.end()
+        );
+    BOOST_CHECK_EQUAL_COLLECTIONS(
+            fst.second.well_rate.begin(), fst.second.well_rate.end(),
+            snd.second.well_rate.begin(), snd.second.well_rate.end()
+        );
+    BOOST_CHECK_EQUAL_COLLECTIONS(
+            fst.second.perf_pressure.begin(), fst.second.perf_pressure.end(),
+            snd.second.perf_pressure.begin(), snd.second.perf_pressure.end()
+        );
+    BOOST_CHECK_EQUAL_COLLECTIONS(
+            fst.second.perf_rate.begin(), fst.second.perf_rate.end(),
+            snd.second.perf_rate.begin(), snd.second.perf_rate.end()
+        );
+}
+
+BOOST_AUTO_TEST_CASE(EclipseReadWriteWellStateData) {
+    test_work_area_type * test_area = test_work_area_alloc("test_Restart");
```

Comment:
This is available as: https://github.com/Ensembles/ert/blob/master/devel/libert_utilxx/include/ert/util/TestArea.hpp